README.md
# platfrom-demo_apps-spaCyNER


changelog.md
v1.2.0
Update Spacy NER API call


js/main.js
let data = {
	retrievedID: [],
};
let state = {};

/* +---------------------+ */
/* | Toggle Popup Alert  | */
/* +---------------------+ */
togglePopUpAlert = (alertTitle, alertMsg) => {
	$('#alertContent').html(alertMsg);
	$('#alertTitle').html(alertTitle);
	$('#alert').modal('toggle');
};

// +--------------------+ //
// | Text area handeler | //
// +--------------------+ //
handelText = (textAreaId) => {
	let text = $(`#${textAreaId}`).val();
	// Calling spaCyNER API
	spaCyNER(text)
		.then((resolve) => {
			// Show result container
			$('#resultContainer').show();
			// Show result container at the side
			$('#functionsContianer').attr('class', 'col-12 col-lg-6');
			// Vertically expand text area
			$('#textArea').attr('rows', 20);
			$('#handelTextBtn, #handelRestartBtn').toggle();
			//console.log(resolve);
		})
		.catch((reject) => {
			// Handle spaCyNER error
			//console.log('Error');
			togglePopUpAlert(
				`Error: ${reject.status}`,
				reject.responseText
					? JSON.parse(reject.responseText).message
					: reject.customErrMsg
			);
		});
};

clearText = (textAreaId) => {
	$(`#${textAreaId}`).val('');
	$('#word-counter').text('0/5000');
};

handelRestart = () => {
	clearText('textArea');
	// Clean rendered CardLinks
	$('#result').empty();
	$('#resultContainer').hide();
	$('#handelTextBtn, #handelRestartBtn').toggle();
	// Reset textarea to fill width
	$('#functionsContianer').attr('class', 'col-12');
	// Reset textarea heighgt
	$('#textArea').attr('rows', 15);
};

// Utility Function, create dom object
createDomElement = (params) => {
	let element = document.createElement(params[0]);
	element.setAttribute('class', params[1]);
	element.setAttribute('style', params[2]);
	return element;
};
// Takes an array of dom element
// Will append element in array order
domAppendHelper = (array) => {
	if (array.length >= 2) {
		for (i = 0; i < array.length - 1; i++) {
			$(array[i]).append(array[i + 1]);
		}
	}
};

// +----------------------------+ //
// |  Render Marsonry CardTags  | //
// +----------------------------+ //

// Example of params input
// Need to return spaCy ner to the format below
let params = {
	// string contains RGB value, to be generated
	icon: 'fa-camera',
	category: 'Some Category',
	details: ['Phase 2', 'Some other category'],
};

// Returns a complete card html component
renderCardTags = (params) => {
	let icon = params.icon;
	let color = randomColor(0, 35, 50, 36, 1, 1, 10);

	let links = '';

	for (item in params.details) {
		let keyword = $.trim(params.details[item]);
		let cardLinkId = `${params.category.replace(
			/[\'\"\`\~\/\>\<\.\,\?\#\$\%\^\&\*\{\}\[\]\:\(\)\s]/g,
			''
		)}-${keyword.replace(
			/[\'\"\`\~\/\>\<\.\,\?\#\$\%\^\&\*\{\}\[\]\:\(\)\s]/g,
			''
		)}-${item}`;

		// Creating individual link item, with hover tool tip
		let wikiLink = `
		<a id="${cardLinkId}" 
		class="card-link icon-link" 
		style="background-color:hsla(${color} .2)"
        >
        ${keyword}
        </a>
        `;
		links += wikiLink;
	}

	let card = `
    <div class="card" style="background-color:hsla(${color} .1)">
        <div class="card-body p-0">

            <div class="card-title-container" 
            style="background-color: hsla(${color} .4); color:#424242 ">
                <span class="mr-2 fas ${icon}">                 
                </span>
                <b clas="m-0 card-title text-center">
                ${params.category}
                </b>
            </div>

            <div class="card-link-container">
                ${links}
            </div>
        </div>
    </div>
    `;

	$('#result').append(card);

	for (item in params.details) {
		let keyword = $.trim(params.details[item]);
		let cardLinkId = `${params.category.replace(
			/[\'\"\`\~\/\>\<\.\,\?\#\$\%\^\&\*\{\}\[\]\:\(\)\s]/g,
			''
		)}-${keyword.replace(
			/[\'\"\`\~\/\>\<\.\,\?\#\$\%\^\&\*\{\}\[\]\:\(\)\s]/g,
			''
		)}-${item}`;

		// Create tool tip
		tippy(`#${cardLinkId}`, {
			content: `
			    <div class='spinner-border spinner-border-sm' role='status'></div>
				<span class='pl-2'>Retrieving data from wikipedia ...</span>`,
			allowHTML: true,
			maxWidth: 500,
			delay: 350,
			// On tool tip triggered
			onShow(instance, partialProps) {
				// If the cardLinkId haven't been retrived
				if (!data.retrievedID.includes(cardLinkId)) {
					// Store retrieved ID, avoid duplicated API call
					data.retrievedID.push(cardLinkId);

					let cardElement = document.getElementById(cardLinkId);
					//console.log(cardElement);
					retrieveKeyword(cardElement).then((content) => {
						instance.setContent(content);
						instance.setProps({
							delay: 200,
							interactive: true,
						});
					}).catch(error =>{
						instance.setContent(error);
						instance.setProps({
							delay: 200,
						});
					});
				}
			},
		});
	}
};

// +--------------------------+ //
// |  Random Color Generator  | //
// +--------------------------+ //
randomColor = (hue, saturate, light, randH, randS, randL, randScale) => {
	// Parametet takes a base H S L value,
	// The random value will be 0 to input value added to base vale
	// RandScalue is for Hue value only, make sure 2 colors are different enough
	let h = hue + Math.floor(Math.random() * randH) * randScale;
	let s = saturate + Math.floor(Math.random() * randS);
	let l = light + Math.floor(Math.random() * randL);
	return `${h}, ${s}%, ${l}%,`;
};

// +----------+ //
// |  Loader  | //
// +----------+ //
let loadingMsg = [
	'Just a moment more, processing your file...',
	'You can buy and sell data securely on Sentient.io’s blockchain network.',
	'Use utility microservices to save time during your app development.',
	'Have a microservice you\'re looking for but can\'t find? Write in to us <a style="text-decoration:underline"  href = "mailto: enquiry@sentient.io">enquiry@sentient.io</a>',
	"Need help with implementing the APIs? Click the 'Help' button at the bottom of the screen to reach out to our support team.",
	'The APIs on our platform are curated carefully to ensure reliability for deployment',
	'Usage discounts are automatically applied as the number of API calls made reaches the next tier',
	'Just a moment more, processing your file...',
];
let loading;
loadingStart = () => {
	$('#loader').toggle();

	let msgIndex = 0;
	loading = window.setInterval(() => {
		$('#loader-text').html(loadingMsg[msgIndex]);
		if (msgIndex < loadingMsg.length) {
			msgIndex += 1;
		} else {
			msgIndex = 1;
		}
	}, 4000);
};
loadingEnd = () => {
	$('#loader').toggle();
	window.clearInterval(loading);
};

$('#textArea').keyup(function () {
	$('#word-counter').text($.trim(this.value.length) + '/5000');
});

/* +-------------------+ */
/* | Handle Text Input | */
/* +-------------------+ */
checkContent = (e) => {
	if (e.target.value == '') {
		$('#handelTextBtn').attr('disabled', '');
	} else {
		$('#handelTextBtn').removeAttr('disabled');
	}
};
document.getElementById('textArea').addEventListener('input', checkContent);


js/spaCyNER.js
spaCyNER = (text) => {
	//console.log('Calling Spacy NER API');
	return new Promise((resolve, reject) => {
		loadingStart();
		$.ajax({
			method: 'post',
			url:
				'https://apis.sentient.io/microservices/nlp/spacyner/v1/getpredictions',
			headers: {
				'x-api-key': apikey,
				'Content-Type': 'application/json',
			},
			data: JSON.stringify({ text: text }),
			success: (response) => {
				let results = response.results;
				loadingEnd();
				//console.log('Success');
				if (Object.keys(results)[0]) {
					for (items in Object.keys(results)) {
						renderCardTags({
							icon: nerMappingFaIcons(Object.keys(results)[items]),
							category: nerMappingCategory(Object.keys(results)[items]),
							details: Object.values(results)[items],
						});
					}
					resolve(results);
				} else {
					reject({
						status: 'No Valid detection',
						customErrMsg:
							'Pleast input longer text with valid English named entities',
					});
				}
			},
			error: (err) => {
				//console.log(`Error: ${err.status}`);
				loadingEnd();
				reject(err);
			},
		});
	});
};

nerMappingCategory = (keyword) => {
	let result;
	let library = {
		person: 'Person',
		norp: 'Nationalities/ Religious/ Political Groups',
		fac: 'Buildings and Facilities',
		org: 'Organisations',
		gpe: 'Countries/ Cities/ States',
		loc: 'Locations',
		product: 'Objects',
		event: 'Events',
		work_of_art: 'Works Of Art',
		law: 'Named Documents Made Into Laws',
		language: 'Language',
		date: 'Date',
		time: 'Time',
		percent: 'Percentage',
		money: 'Monetary Values',
		quantity: 'Measurements',
		ordinal: 'Ordinal',
		cardinal: 'Numbers',
	};
	result = library[keyword] != undefined ? library[keyword] : keyword;
	return result;
};

nerMappingFaIcons = (keyword) => {
	let result;
	let library = {
		person: 'fa-users',
		norp: 'fa-globe',
		fac: 'fa-hotel',
		org: 'fa-university',
		gpe: 'fa-flag',
		loc: 'fa-map-marked',
		product: 'fa-cubes',
		event: 'fa-calendar-week',
		work_of_art: 'fa-book',
		law: 'fa-gavel',
		language: 'fa-language',
		date: 'fa-calendar-day',
		time: 'fa-stopwatch',
		percent: 'fa-percent',
		money: 'fa-money-bill-wave',
		quantity: 'fa-ruler',
		ordinal: 'fa-tag',
		cardinal: 'fa-tags',
	};
	result = library[keyword] != undefined ? library[keyword] : keyword;
	return result;
};


js/wikipediaRetrieval.js
var RETRIEVE_KEYWORD_TIMER;

wikipediaRetrieval = (keyword) => {
	return new Promise((resolve, reject) => {
		//console.log('Calling Wikipedia Retrieval');
		//console.log(keyword.trim());
		const data = JSON.stringify({
			keyword: keyword.trim(),
		});

		const xhr = new XMLHttpRequest();

		xhr.addEventListener('readystatechange', function () {
			if (this.readyState === this.DONE) {
				if (JSON.parse(this.response).results) {
					//console.log('Wiki Retrieval Success');
					//console.log(this.response);
					resolve(this.response);
				} else {
					reject(this.response);
				}
			}
		});

		xhr.open(
			'POST',
			'https://apis.sentient.io/microservices/utility/wikipedia/v1/getresults'
		);
		xhr.setRequestHeader('content-type', 'application/json');
		xhr.setRequestHeader('x-api-key', apikey);

		xhr.send(data);
	});
};

delayRetrieveKeyword = (e) => {
	RETRIEVE_KEYWORD_TIMER = setTimeout(function () {
		//console.log('Retrieve keyword timer - Start');
		//console.log(e.id);
		retrieveKeyword(e);
	}, 350);
};

stopWikiRetrieveKeyword = function () {
	clearTimeout(RETRIEVE_KEYWORD_TIMER);
	//console.log('Retrieve keyword timer - Stopped');
};

retrieveKeyword = (e) => {
	return new Promise((resolve, reject) => {
		//console.log('Retrieve Keyword - Start');
		let keyword = e.innerHTML;
		let id = e.id;
		// Add loading icon
		$(`#${id}`).addClass('icon-loading');

		wikipediaRetrieval(keyword)
			.then((res) => {
				let response = JSON.parse(res);

				$(`#${id}`).removeClass('icon-loading');
				$(`#${id}`).addClass('icon-check');
				//console.log(res);

				let thumbnail = '';
				if (response.results.thumbnail) {
					thumbnail = `
								<img class="m-3 border" width="120" src="${response.results.thumbnail.source}" />
								`;
				}

				let content = '';

				if (
					response.message ===
					'Disambigutous keyword.Please choose from the below list and try again'
				) {
					content = '<p class="pb-1">Disambiguation keyword lists: </p>';
					response.results.forEach((keyword) => {
						content += `<p class="p-0 m-0">- ${keyword}</p>`;
					});
				} else {
					content = `
					<div class="m-3 d-flex flex-row justify-content-between">
		
					<div class="col-6 p-0 m-0 text-left">
					<h5>${keyword}</h5>
					
					<p><b>Page ID: </b> ${response.results.pageid}</p>
					<b class="mt-1">Wikipedia Page URL:</b>
					<a class="s-link mr-2" href="${response.results.url}" target="_blank">
					${response.results.url}
					</a>
					</div>
					<br/>
					<div class="col-6 p-0 m-0 text-right">
					${thumbnail}
					</div>
					</div>
		
					<div class="col-12 p-3 text-left">
					<b>Summary:</b>
					<p>${response.results.summary}</p>
					</div>
					`;
				}
				resolve(content);
			})
			.catch((err) => {
				//console.log(err);

				$(`#${id}`).removeClass('icon-loading');
				$(`#${id}`).addClass('icon-no-result');

				reject(JSON.parse(err).message);
			});
	});
};


js/OLD_main.js
let data = {};
let state = {};

/* +---------------------+ */
/* | Toggle Popup Alert  | */
/* +---------------------+ */
togglePopUpAlert = (alertTitle, alertMsg) => {
	$('#alertContent').html(alertMsg);
	$('#alertTitle').html(alertTitle);
	$('#alert').modal('toggle');
};

// +--------------------+ //
// | Text area handeler | //
// +--------------------+ //
handelText = (textAreaId) => {
	let text = $(`#${textAreaId}`).val();
	// Calling spaCyNER API
	spaCyNER(text)
		.then((resolve) => {
			// Show result container
			$('#resultContainer').show();
			// Show result container at the side
			$('#functionsContianer').attr('class', 'col-12 col-lg-6');
			// Vertically expand text area
			$('#textArea').attr('rows', 20);
			$('#handelTextBtn, #handelRestartBtn').toggle();
			console.log(resolve);
		})
		.catch((reject) => {
			// Handle spaCyNER error
			console.log('Error');
			togglePopUpAlert(
				`Error: ${reject.status}`,
				reject.responseText
					? JSON.parse(reject.responseText).message
					: reject.customErrMsg
			);
		});
};

clearText = (textAreaId) => {
	$(`#${textAreaId}`).val('');
	$('#word-counter').text('0/5000');
};

handelRestart = () => {
	clearText('textArea');
	// Clean rendered CardLinks
	$('#result').empty();
	$('#resultContainer').hide();
	$('#handelTextBtn, #handelRestartBtn').toggle();
	// Reset textarea to fill width
	$('#functionsContianer').attr('class', 'col-12');
	// Reset textarea heighgt
	$('#textArea').attr('rows', 15);
};

// Utility Function, create dom object
createDomElement = (params) => {
	let element = document.createElement(params[0]);
	element.setAttribute('class', params[1]);
	element.setAttribute('style', params[2]);
	return element;
};
// Takes an array of dom element
// Will append element in array order
domAppendHelper = (array) => {
	if (array.length >= 2) {
		for (i = 0; i < array.length - 1; i++) {
			$(array[i]).append(array[i + 1]);
		}
	}
};

// +----------------------------+ //
// |  Render Marsonry CardTags  | //
// +----------------------------+ //

// Example of params input
// Need to return spaCy ner to the format below
let params = {
	// string contains RGB value, to be generated
	icon: 'fa-camera',
	category: 'Some Category',
	details: ['Phase 2', 'Some other category'],
};

// Returns a complete card html component
renderCardTags = (params) => {
	let icon = params.icon;
	let color = randomColor(0, 35, 50, 36, 1, 1, 10);
	let card = createDomElement([
		'div',
		'card',
		`background-color: hsla(${color} .1)`,
	]);
	let cardBody = createDomElement(['div', 'card-body p-0', '']);
	let cardTitleContainer = createDomElement([
		'div',
		'card-title-container',
		`background-color: hsla(${color} .4); color:#424242 `,
	]);
	let titleIcon = createDomElement(['span', `mr-2 fas ${icon}`, '']);
	let title = createDomElement(['b', 'm-0 card-title text-center', '']);
	let cardLinkContainer = createDomElement(['div', 'card-link-container', '']);
	$(title).html(`${params.category}`);

	$(cardTitleContainer).append(titleIcon, title);
	$(cardBody).append(cardTitleContainer, cardLinkContainer);
	domAppendHelper(['#result', card, cardBody]);

	for (item in params.details) {
		let keyword = $.trim(params.details[item]);

		// Creating individual link item, with hover tool tip
		let wikiLink = createDomElement([
			'a',
			'card-link icon-link',
			`background-color: hsla(${color} .2)`,
		]);
		let cardLinkId = `${params.category.replace(
			/[\(\)\s\'\/]/g,
			''
		)}-${keyword.replace(/[\(\)\s\'\/]/g, '')}-${item}`;

		$(wikiLink).html(keyword);

		//Hover to fetch data from wikipedia
		// $(wikiLink).attr(
		// 	'onmouseenter',
		// 	`delayWikiRetrieval("${keyword}", "${cardLinkId}")`
		// );

		// $(wikiLink).attr('onmouseout', `cancelWikiRetrival()`);
		$(wikiLink).attr('onClick', `wikiRetrieveKeyword(this)`);

		// Create bootstrap tool tip
		$(wikiLink).attr('data-toggle', 'tooltip');
		$(wikiLink).attr('data-html', 'true');
		$(wikiLink).css('text-decoration', 'none !important');
		//$(wikiLink).attr('href', '#');
		// $(wikiLink).attr('title', 'Click keyword to retrieve data from wikipedia');
		$(wikiLink).attr('id', `${cardLinkId}`);
		$(cardLinkContainer).append(wikiLink);
	}
	// Call the function below everytime when there is a DOM change to display the Bootstrap tool tip
	// $('[data-toggle="tooltip"]').tooltip();
};

// +----------------------------------+ //
// |  Retrive Keyword from wikipedia  | //
// +----------------------------------+ //
wikiRetrieveKeyword = (e) => {
	// Get text content of clicked link
	let keyword = e.innerHTML;
	wikipediaRetrieval(keyword)
		.then((result) => {
			console.log(result);
		})
		.catch((err) => {
			console.log(err);
		});
};

// +--------------------------+ //
// |  Random Color Generator  | //
// +--------------------------+ //
randomColor = (hue, saturate, light, randH, randS, randL, randScale) => {
	// Parametet takes a base H S L value,
	// The random value will be 0 to input value added to base vale
	// RandScalue is for Hue value only, make sure 2 colors are different enough
	let h = hue + Math.floor(Math.random() * randH) * randScale;
	let s = saturate + Math.floor(Math.random() * randS);
	let l = light + Math.floor(Math.random() * randL);
	return `${h}, ${s}%, ${l}%,`;
};

// +----------+ //
// |  Loader  | //
// +----------+ //
let loadingMsg = [
	'Just a moment more, processing your file...',
	'You can buy and sell data securely on Sentient.io’s blockchain network.',
	'Use utility microservices to save time during your app development.',
	'Have a microservice you\'re looking for but can\'t find? Write in to us <a style="text-decoration:underline"  href = "mailto: enquiry@sentient.io">enquiry@sentient.io</a>',
	"Need help with implementing the APIs? Click the 'Help' button at the bottom of the screen to reach out to our support team.",
	'The APIs on our platform are curated carefully to ensure reliability for deployment',
	'Usage discounts are automatically applied as the number of API calls made reaches the next tier',
	'Just a moment more, processing your file...',
];
let loading;
loadingStart = () => {
	$('#loader').toggle();

	let msgIndex = 0;
	loading = window.setInterval(() => {
		$('#loader-text').html(loadingMsg[msgIndex]);
		if (msgIndex < loadingMsg.length) {
			msgIndex += 1;
		} else {
			msgIndex = 1;
		}
	}, 4000);
};
loadingEnd = () => {
	$('#loader').toggle();
	window.clearInterval(loading);
};

$('#textArea').keyup(function () {
	$('#word-counter').text($.trim(this.value.length) + '/5000');
});

// +---------+ //
// |  Alert  | //
// +---------+ //
// toggleAlert = (data) => {
// 	console.log('Alert toggled');
// 	let id = data.id;
// 	let message = data.message;
// 	let color = data.color;

// 	let alertContainer = document.createElement('div');
// 	alertContainer.setAttribute('id', `${id}-alert`);
// 	alertContainer.setAttribute('class', 'dmo-alert-container');
// 	let alertContent = document.createElement('div');
// 	alertContent.setAttribute('class', 'dmo-alert-content');
// 	let alertText = document.createElement('span');
// 	alertText.setAttribute('style', `color:${color}`);
// 	alertText.innerHTML = message;

// 	domAppendHelper(['#dmo-alert', alertContainer, alertContent, alertText]);
// 	//$('#dmo-alert').append(alertContainer)

// 	setTimeout(() => {
// 		console.log('Removing Alert');
// 		$(`${id}-alert`).remove();
// 	}, 5000);
// };

/* +-------------------+ */
/* | Handle Text Input | */
/* +-------------------+ */
checkContent = (e) => {
	if (e.target.value == '') {
		$('#handelTextBtn').attr('disabled', '');
	} else {
		$('#handelTextBtn').removeAttr('disabled');
	}
};
document.getElementById('textArea').addEventListener('input', checkContent);
